### PR TITLE
[SPARK-18834][SS] Expose event time stats through StreamingQueryProgress

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

- Changed `StreamingQueryProgress.watermark` to `StreamingQueryProgress.queryTimestamps` which is a `Map[String, String]` containing the following keys: "eventTime.max", "eventTime.min", "eventTime.avg", "processingTime", "watermark". All of them UTC formatted strings.

- Renamed `StreamingQuery.timestamp` to `StreamingQueryProgress.triggerTimestamp` to differentiate from `queryTimestamps`. It has the timestamp of when the trigger was started. 

## How was this patch tested?

Updated tests

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryStatusAndProgressSuite.scala [hunk 1]
```diff
@@ -38,13 +38,18 @@ class StreamingQueryStatusAndProgressSuite extends SparkFunSuite {
...
-        |  "timestamp" : "2016-12-05T20:54:20.827Z",
+        |  "triggerTimestamp" : "2016-12-05T20:54:20.827Z",
         |  "numInputRows" : 678,
         |  "inputRowsPerSecond" : 10.0,
         |  "durationMs" : {
         |    "total" : 0
         |  },
-        |  "currentWatermark" : 3,
+        |  "queryTimestamps" : {
+        |    "eventTime.avg" : "2016-12-05T20:54:20.827Z",
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryStatusAndProgressSuite.scala]: @marmbrus the dot is in the key. when parsing using our json DF, we may have to refer to the column as "progress.queryTimestamps.`eventTim.max`"

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryStatusAndProgressSuite.scala [hunk 2]
```diff
@@ -38,13 +38,18 @@ class StreamingQueryStatusAndProgressSuite extends SparkFunSuite {
         |  "id" : "${testProgress1.id.toString}",
         |  "runId" : "${testProgress1.runId.toString}",
         |  "name" : "myName",
-        |  "timestamp" : "2016-12-05T20:54:20.827Z",
+        |  "triggerTimestamp" : "2016-12-05T20:54:20.827Z",
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryStatusAndProgressSuite.scala]: @marmbrus One idea is to not have a top level timestamp, and merge this with the `queryTimestamps`. Then this would be a key `triggerStartTime` in the `queryTimestamps` map. In fact then we can rename `queryTimestamps` to `timestamps`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 3]
```diff
@@ -360,6 +360,24 @@ class StreamExecution(
     if (hasNewData) {
       // Current batch timestamp in milliseconds
       offsetSeqMetadata.batchTimestampMs = triggerClock.getTimeMillis()
+      // Update the eventTime watermark if we find one in the plan.
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: I moved this such that the watermark is updated before starting a batch, rather than after finishing a batch. This keeps batchWatermarkMs consistent with batchTimestampsMs, both are set before starting a batch, and reduces complexities in the ProgressReporter.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryStatusAndProgressSuite.scala (hunk 1 above)
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryStatusAndProgressSuite.scala]: One option is we could use dashes instead, but honestly I'm getting a little confused by all these timestamps.  In particular, I'm not sure what is the difference between `triggerTimestamp` and `processingTime` (and, I think that having `processingTime` mean anything different than `System.currentTimeMillis` will be confusing to users coming from other systems).

The two things that I think you really want to be able to track from the metrics feed are:
 - the actual timestamp when this progress update was produced.  I think this should remain top level and be called `timestamp`.
 - Stats about the event time, so that I can know roughly what data is present in this batch.  This can be useful for several reasons, including finding other problems upstream.  The more I think about this, I think you just want to see this as `"eventTime": { "min": ..., "max": ..., "watermark": ... }`

Its actually not clear to me why you need to know the original batch start time.  In 99% of cases this is the same as the `timestamp`.  If you are executing, a batch due to a failure, it will be different.  But, why does an outside monitoring job care?  I can't come up with any interesting graphs that you would make with this extra field.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala (hunk 3 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: +1, this makes sense.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryStatusAndProgressSuite.scala (hunk 1 above)
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryStatusAndProgressSuite.scala]: It is available, it is stored in a human readable format in the offset log (BTW, is that a long or a timestamp?).  I think in the log run we'll want to open up another API that gives you access to this log, but I think that can come later.  For now, its still pretty easy to find.

Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryStatusAndProgressSuite.scala]: But the log gets cleaned up continuously, so will not be available if you are trying to debug it a day or an hour later.

Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryStatusAndProgressSuite.scala]: Not any more... we keep 1000 now, right? This just really feels like we are reaching for a use case.  We can always add it, but I think the way it is done currently in this PR is very confusing and having `timestamp` and `eventTime` are very clear.

Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryStatusAndProgressSuite.scala]: 1000 *  5 seconds batch is < 2 hours. Anyways, I think it will be equally or more confusing adding it later as a top-level field in Progress (e.g. `timestamp`, `eventTime`, `processingTime` ?). It may be better to have the top-level `timestamp`, and all other execution level detailed timestamps inside a single map. 

Anyways, I am updating the PR, but I think this is a little short-sighted.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala [hunk 4]
```diff
@@ -33,27 +34,6 @@ import org.apache.spark.sql.catalyst.util.DateTimeUtils
 
 /**
  * :: Experimental ::
- * Information about updates made to stateful operators in a [[StreamingQuery]] during a trigger.
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala]: I moved this below to keep the StreamingQueryProgress class first in the file. More important code first.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala]: Thats actually the opposite of how most code in SQL is laid out, so I think it would be better to avoid this change.  The logic here is declarations that are use later should come first (references before declaration make it harder to read), and stuff at the end of the file is kind of hidden.

Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala]: aah, then for consistency, SourceProgress and SinkProgress should also be before StreamingQueryProgress. But thats a bigger change should be done in a different PR.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryStatusAndProgressSuite.scala [hunk 5]
```diff
@@ -44,7 +44,12 @@ class StreamingQueryStatusAndProgressSuite extends SparkFunSuite {
         |  "durationMs" : {
         |    "total" : 0
         |  },
-        |  "currentWatermark" : 3,
+        |  "eventTime" : {
```
Comment (brkyvz) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryStatusAndProgressSuite.scala]: awesome!

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala [hunk 6]
```diff
@@ -67,7 +67,7 @@ class StateOperatorProgress private[sql](
  *                Similarly, when there is no data to be processed, the batchId will not be
  *                incremented.
  * @param durationMs The amount of time taken to perform various operations in milliseconds.
- * @param currentWatermark The current event time watermark in milliseconds
+ * @param eventTime Statistics of event time seen in this batch
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala]: nit: since you are touching this file, could you also fix the comment of `@param timestamp`? It's better to document the format as well, such as, `The beginning time of the trigger in ISO8601 format. (e.g., 2016-12-05T20:54:20.827Z)`

Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala]: In addition, could you also add an example in the comment of `eventTime`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/EventTimeWatermarkExec.scala [hunk 7]
```diff
@@ -28,24 +26,48 @@ import org.apache.spark.sql.types.MetadataBuilder
...
+  }
+
+  def merge(that: EventTimeStats): Unit = {
+    this.max = math.max(this.max, that.max)
+    this.min = math.min(this.min, that.min)
+    this.sum += that.sum
+    this.count += that.count
+  }
+
+  def avg: Long = (sum.toDouble / count).toLong
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/EventTimeWatermarkExec.scala]: nit: why not use `sum / count`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/EventTimeWatermarkExec.scala [hunk 8]
```diff
@@ -61,15 +83,14 @@ case class EventTimeWatermarkExec(
...
-  sparkContext.register(maxEventTime)
+  val eventTimeStats = new EventTimeStatsAccum()
+  sparkContext.register(eventTimeStats)
 
   override protected def doExecute(): RDD[InternalRow] = {
     child.execute().mapPartitions { iter =>
       val getEventTime = UnsafeProjection.create(eventTime :: Nil, child.output)
       iter.map { row =>
-        maxEventTime.add(getEventTime(row).getLong(0))
+        eventTimeStats.add((getEventTime(row).getLong(0).toDouble / 1000).toLong)
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/EventTimeWatermarkExec.scala]: nit: could be `getEventTime(row).getLong(0) / 1000`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/EventTimeWatermarkExec.scala [hunk 9]
```diff
@@ -28,24 +26,48 @@ import org.apache.spark.sql.types.MetadataBuilder
 import org.apache.spark.unsafe.types.CalendarInterval
 import org.apache.spark.util.AccumulatorV2
 
-/** Tracks the maximum positive long seen. */
-class MaxLong(protected var currentValue: Long = 0)
-  extends AccumulatorV2[Long, Long] {
+/** Class for collecting event time stats with an accumulator */
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/EventTimeWatermarkExec.scala]: nit: please document the time unit.

**Review by zsxwing — COMMENTED** [review]
Looks good overall. Just some nits.

**Comment by zsxwing** [comment]
LGTM

**Comment by marmbrus** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/EventTimeWatermarkExec.scala (hunk 9 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/EventTimeWatermarkExec.scala]: Documented it in EventTimeWatermarkExec

**Comment by tdas** [comment]
Merging to master and 2.1

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala [hunk 10]
```diff
@@ -61,13 +61,20 @@ class StateOperatorProgress private[sql](
...
+ * @param timestamp Beginning time of the trigger in ISO8601 format, i.e. UTC timestamps.
  * @param batchId A unique id for the current batch of data being processed.  Note that in the
  *                case of retries after a failure a given batchId my be executed more than once.
  *                Similarly, when there is no data to be processed, the batchId will not be
  *                incremented.
  * @param durationMs The amount of time taken to perform various operations in milliseconds.
- * @param currentWatermark The current event time watermark in milliseconds
+ * @param eventTime Statistics of event time seen in this batch. It may contain the following keys:
+ *                 {
+ *                   "max" -> "2016-12-05T20:54:20.827Z"  // maximum event time seen in this trigger
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala]: Hi all, I am just leaving a comment as a gentle reminder to note that we probably should replace `<` or `>` to other ones such as `{@literal <}` or `{@literal >}` in the future. Please refer https://github.com/apache/spark/pull/16013#discussion_r89666914. This causes javadoc8 break.

```
[error] .../java/org/apache/spark/sql/streaming/StreamingQueryProgress.java:19: error: bad use of '>'
[error]  *                   "max" -> "2016-12-05T20:54:20.827Z"  // maximum event time seen in this trigger
[error]                             ^
[error] .../java/org/apache/spark/sql/streaming/StreamingQueryProgress.java:20: error: bad use of '>'
[error]  *                   "min" -> "2016-12-05T20:54:20.827Z"  // minimum event time seen in this trigger
[error]                             ^
[error] .../java/org/apache/spark/sql/streaming/StreamingQueryProgress.java:21: error: bad use of '>'
[error]  *                   "avg" -> "2016-12-05T20:54:20.827Z"  // average event time seen in this trigger
[error]                             ^
[error] .../java/org/apache/spark/sql/streaming/StreamingQueryProgress.java:22: error: bad use of '>'
[error]  *                   "watermark" -> "2016-12-05T20:54:20.827Z"  // watermark used in this trigger
[error]                                   ^
```
